update join hint & add missing table functions (#209)
kuzudb/kuzu-docs: src/content/docs/developer-guide/hint-join-order.md

## Join order hint

Consider the following query
```cypher
MATCH (a:person)-[e:LivesIn]->(b:City)
RETURN *;
```

> The system optimizer has its freedom to generate a query plan that find all edges from `a` to `b` using forward adjcency list or from `b` to `a` using backward adjcency list. Both query plan are semantically equivalent and generate the same query result. There is, however, a potential performance difference because `a` and `b` may have different cardinalities.

Reviewer: ```suggestion
The system optimizer has its freedom to generate a query plan that finds all edges from `a` to `b` using forward adjacency list or from `b` to `a` using a backward adjacency list. Both query plans are semantically equivalent and generate the same query result. There is, however, a potential performance difference because `a` and `b` may have different cardinalities.
```

> To enforce a particular join order (you may want to do this for experimental purpose or the optimizer gives a sub-optimal plan), use `HINT` and write a join order with the following rules

Reviewer: ```suggestion
To enforce a particular join order (you may want to do this for experimental purposes, or when the optimizer provides a sub-optimal plan), use `HINT` and write a join order with the following rules:
```

> - all node/relationship variable must be named and appear in the join order

Reviewer: ```suggestion
- All node/relationship variable must be named and appear in the join order
```

> - the join order is a binary tree where tree node is either a variable or a sub join order tree enclosed by parentheses

Reviewer: ```suggestion
- The join order is a binary tree where the tree node is either a variable or a sub-join order tree enclosed in parentheses
```

> E.g. we can enforce a query plan that scans from `b` with backward adjacency list with the following

Reviewer: ```suggestion
For example, we can enforce a query plan that scans from `b` with a backward adjacency list as follows:
```

```cypher
MATCH (a)-[e]->(b)
WHERE a.ID = 0
HINT a JOIN (e JOIN b)
RETURN *;
```

### Multi-way join hint

> This section assumes you have the knowledge about worst-case-optimal join. For cyclic queries, we allow not only binary join hint but also multi-way join (worst-case-optimal-join) hint. When hinting k-way joins, the following rules needs to be satisfied

Reviewer: ```suggestion
This section assumes you are knowledgeable about worst-case-optimal joins (WCOJ). For cyclic queries, we allow not only binary join hints but also multi-way join (worst-case-optimal-join) hints. When hinting k-way joins, the following rules needs to be satisfied:
```

> - use keyword `MULTI_JOIN` instead of `JOIN`.

Reviewer: ```suggestion
- Use the keyword `MULTI_JOIN` instead of `JOIN`.
```

> - within `k` tree nodes, `k-1` of them must be edge variables these edges must be intersected at a single node.

Reviewer: ```suggestion
- Within `k` tree nodes, `k-1` of them must be edge variables - these edges must intersect at a single node.
```

When we perform the k-way join between a subplan and `k-1` edges, for each input tuple from the subplan, we find an adjacency list from each edge table and sort-merge join them on the intersected node.

> E.g. the following query, find forward adjacency list for `e2` and `e3` and intersect them over `c` for each `(a, e1, b)` input.

Reviewer: ```suggestion
For example, in the following query, we find the forward adjacency list for `e2` and `e3` and intersect them over `c` for each `(a, e1, b)` input.
```
